### ISSUE TITLE
Documented but unknown flags "--name" and "--target-cpu-percentile" for recommender

### ISSUE THREAD

**Issue by dmlabs** [kubernetes/autoscaler#5451]
**Which component are you using?**:  
vertical-pod-autoscaler

**What version of the component are you using?**:  
Component version: 0.12.0

**What k8s version are you using (`kubectl version`)?**:  
$ kubectl version  
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.2", GitCommit:"9d142434e3af351a628bffee3939e64c681afa4d", GitTreeState:"clean", BuildDate:"2022-01-19T17:35:46Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.8+rke2r1", GitCommit:"fdc77503e954d1ee641c0e350481f7528e8d068b", GitTreeState:"clean", BuildDate:"2022-11-10T17:56:22Z", GoVersion:"go1.18.8b7", Compiler:"gc", Platform:"linux/amd64"}
```

**What environment is this in?**:  
Rancher RKE2 Cluster on-prem

**What did you expect to happen?**:  
Deploy an additional recommender with the flags `--name=frugal` and `--target-cpu-percentile=0.50` as documented here: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#starting-multiple-recommenders

I expected the additional recommender to start and to be available.

**What happened instead?**:  
The pod of the additional recommender does not start (CrashLoopBackOff) and the logs say that "--name" and "--target-cpu-percentile" are unknown flags. Complete logs:
```
unknown flag: --name
unknown flag: --target-cpu-percentile
Usage of /recommender:
--add-dir-header If true, adds the file directory to the header of the log messages
--address string The address to expose Prometheus metrics. (default ":8942")
--alsologtostderr log to standard error as well as files (no effect when -logtostderr=true)
--checkpoints-gc-interval duration How often orphaned checkpoints should be garbage collected (default 10m0s)
--checkpoints-timeout duration Timeout for writing checkpoints since the start of the recommender's main loop (default 1m0s)
--container-name-label string Label name to look for container names (default "name")
--container-namespace-label string Label name to look for container namespaces (default "namespace")
--container-pod-name-label string Label name to look for container pod names (default "pod_name")
--cpu-histogram-decay-half-life duration The amount of time it takes a historical CPU usage sample to lose half of its weight. (default 24h0m0s)
--history-length string How much time back prometheus have to be queried to get historical metrics (default "8d")
--history-resolution string Resolution at which Prometheus is queried for historical metrics (default "1h")
--kube-api-burst float QPS burst limit when making requests to Kubernetes apiserver (default 10)
--kube-api-qps float QPS limit when making requests to Kubernetes apiserver (default 5)
--kubeconfig string Path to a kubeconfig. Only required if out-of-cluster.
--log-backtrace-at traceLocation when logging hits line file:N, emit a stack trace (default :0)
--log-dir string If non-empty, write log files in this directory (no effect when -logtostderr=true)
--log-file string If non-empty, use this log file (no effect when -logtostderr=true)
--log-file-max-size uint Defines the maximum size a log file can grow to (no effect when -logtostderr=true). Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
--logtostderr log to standard error instead of files (default true)
--memory-aggregation-interval duration The length of a single interval, for which the peak memory usage is computed. Memory usage peaks are aggregated in multiples of this interval. In other words there is one memory usage sample per interval (the maximum usage over that interval) (default 24h0m0s)
--memory-aggregation-interval-count int The number of consecutive memory-aggregation-intervals which make up the MemoryAggregationWindowLength which in turn is the period for memory usage aggregation by VPA. In other words, MemoryAggregationWindowLength = memory-aggregation-interval * memory-aggregation-interval-count. (default 8)
--memory-histogram-decay-half-life duration The amount of time it takes a historical memory usage sample to lose half of its weight. In other words, a fresh usage sample is twice as 'important' as one with age equal to the half life period. (default 24h0m0s)
--memory-saver If true, only track pods which have an associated VPA
--metric-for-pod-labels string Which metric to look for pod labels in metrics (default "up{job=\"kubernetes-pods\"}")
--min-checkpoints int Minimum number of checkpoints to write per recommender's main loop (default 10)
--one-output If true, only write logs to their native severity level (vs also writing to each lower severity level; no effect when -logtostderr=true)
--pod-label-prefix string Which prefix to look for pod labels in metrics (default "pod_label_")
--pod-name-label string Label name to look for pod names (default "kubernetes_pod_name")
--pod-namespace-label string Label name to look for pod namespaces (default "kubernetes_namespace")
--pod-recommendation-min-cpu-millicores float Minimum CPU recommendation for a pod (default 25)
--pod-recommendation-min-memory-mb float Minimum memory recommendation for a pod (default 250)
--prometheus-address string Where to reach for Prometheus metrics
--prometheus-cadvisor-job-name string Name of the prometheus job name which scrapes the cAdvisor metrics (default "kubernetes-cadvisor")
--prometheus-query-timeout string How long to wait before killing long queries (default "5m")
--recommendation-margin-fraction float Fraction of usage added as the safety margin to the recommended request (default 0.15)
--recommender-interval duration How often metrics should be fetched (default 1m0s)
--skip-headers If true, avoid header prefixes in the log messages
--skip-log-headers If true, avoid headers when opening log files (no effect when -logtostderr=true)
--stderrthreshold severity logs at or above this threshold go to stderr when writing to files and stderr (no effect when -logtostderr=true or -alsologtostderr=false) (default 2)
--storage string Specifies storage mode. Supported values: prometheus, checkpoint (default)
-v, --v Level number for the log level verbosity
--vmodule moduleSpec comma-separated list of pattern=N settings for file-filtered logging
--vpa-object-namespace string Namespace to search for VPA objects and pod stats. Empty means all namespaces will be used.
```

**How to reproduce it (as minimally and precisely as possible)**:  
1. Install VPA as documented: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler#install-command
2. Deploy an additional recommender, e. g. https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/deploy/recommender-deployment-low.yaml

Can you please fix this issue? I am happy to help with the testing afterwards.

**Comment by Shubham82** [comment]
Hi @dmlabs
I tried this one, it is working if we remove the double hyphen sign `--` from the front of `target-cpu-percentile` and `name` arguments.
instead of using:
 ```
args:
   - --name=frugal
   - --target-cpu-percentile=0.50
```
     
we have to use the following:
 instead of using:
 ```
args:
   - name=frugal
   - target-cpu-percentile=0.50
```

**Comment by Shubham82** [comment]
Updated Yaml which I used for this, Please find the observation below:

```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: vpa-recommender
  namespace: kube-system
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: vpa-recommender-performance
  namespace: kube-system
spec:
  replicas: 1
  selector:
    matchLabels:
      app: vpa-recommender-performance
  template:
    metadata:
      labels:
        app: vpa-recommender-performance
    spec:
      serviceAccountName: vpa-recommender
      securityContext:
        runAsNonRoot: true
        runAsUser: 65534 # nobody
      containers:
      - name: recommender
        image: k8s.gcr.io/autoscaling/vpa-recommender:0.12.0
        imagePullPolicy: Always
        args:
          - target-cpu-percentile=0.50
          - name=frugal
        resources:
          limits:
            cpu: 200m
            memory: 1000Mi
          requests:
            cpu: 50m
            memory: 500Mi
        ports:
        - name: prometheus
          containerPort: 8942

```
```
$ kubectl get pod -n kube-system
NAME                                         READY   STATUS    RESTARTS      AGE
coredns-565d847f94-8vqfw                     1/1     Running   0             61m
etcd-minikube-vpa-1.12                       1/1     Running   0             61m
kube-apiserver-minikube-vpa-1.12             1/1     Running   0             61m
kube-controller-manager-minikube-vpa-1.12    1/1     Running   0             61m
kube-proxy-lvk24                             1/1     Running   0             61m
kube-scheduler-minikube-vpa-1.12             1/1     Running   0             61m
metrics-server-769cd898cd-29tsh              1/1     Running   0             58m
storage-provisioner                          1/1     Running   1 (60m ago)   61m
vpa-admission-controller-64cb9cfb4b-6wf46    1/1     Running   0             55m
vpa-recommender-6fdd69df6c-2j7k8             1/1     Running   0             55m
vpa-recommender-performance-b7c648f6-gxnhn   1/1     Running   0             4m10s
vpa-updater-8567bff646-rctpl                 1/1     Running   0             55m

```
I tried this on minikube having k8s version v1.25.3

**Comment by Shubham82** [comment]
could you please check on your system, if it will work for you then I will correct this.

**Comment by voelzmo** [comment]
This flag doesn't exist in the released vpa 0.12. Compare https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-0.12.0/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go and https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/recommender/logic/recommender.go

The PR was merged on Sept 26th while the 0.12 release was cut on Sept 20th. Note that release notes for 0.12 also don't mention this change: https://github.com/kubernetes/autoscaler/releases/tag/vertical-pod-autoscaler-0.12.0 and @jbartosik indicated this was intentionally not included in the release: https://github.com/kubernetes/autoscaler/pull/4799#pullrequestreview-1094847196

TL;DR: you'll have to wait for vpa 0.13 to be released, which will be somewhat soon: https://github.com/kubernetes/autoscaler/issues/5355

**Comment by jbartosik** [comment]
I think @voelzmo answered the issue, I'm closing it.

Please reopen if you don't think we should close it yet

**Comment by Littlericket** [comment]
@jbartosik the arg --name still doenst apply to the recommender, as defined here: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/deploy/recommender-deployment-high.yaml#L32C31-L32C31

**Comment by jbartosik** [comment]
@Littlericket thanks, I sent a PR to fix the documentation

**Comment by Shubham82** [comment]
@jbartosik we have to backport #6081 to the vpa-release-0.13 and vpa-release-0.14 branches.
I will open PR for it soon.

**Comment by Shubham82** [comment]
Raised the following PR to fix this issue:
For VPA  0.13 #6082
For VPA  0.14 #6083
